File: ScapesCore/src/main/java/org/tobi29/scapes/client/input/touch/GuiControllerTouch.java
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.client.input.touch;

import java8.util.Optional;
import java8.util.stream.Stream;
import org.tobi29.scapes.engine.ScapesEngine;
import org.tobi29.scapes.engine.gui.*;
import org.tobi29.scapes.engine.input.ControllerBasic;
import org.tobi29.scapes.engine.input.ControllerTouch;
import org.tobi29.scapes.engine.utils.Pair;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.math.vector.MutableVector2;
import org.tobi29.scapes.engine.utils.math.vector.Vector2;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class GuiControllerTouch extends GuiController {
    private final ControllerTouch controller;
    private final Map<ControllerTouch.Tracker, Finger> fingers =
            new ConcurrentHashMap<>();
    private List<Pair<GuiCursor, ControllerBasic.PressEvent>> clicks =
            Collections.emptyList();

    public GuiControllerTouch(ScapesEngine engine, ControllerTouch controller) {
        super(engine);
        this.controller = controller;
    }

    @Override
    public void update(double delta) {
        double ratio = 540.0 / engine.container().containerHeight();
        List<Pair<GuiCursor, ControllerBasic.PressEvent>> newClicks =
                new ArrayList<>();
        Map<ControllerTouch.Tracker, Finger> newFingers =
                new ConcurrentHashMap<>();
        controller.fingers().forEach(tracker -> {
            Finger fetch = fingers.get(tracker);
            if (fetch == null) {
                fetch = new Finger(tracker.pos());
                Finger finger = fetch;
                handleFinger(finger, ratio);
                finger.dragX = finger.cursor.guiX();
                finger.dragY = finger.cursor.guiY();
                engine.guiStack().fireEvent(
                        new GuiComponentEvent(finger.cursor.guiX(),
                                finger.cursor.guiY()),
                        (component, event, engine) -> true, engine)
                        .ifPresent(component -> {
                            if (!Streams.of(fingers.values())
                                    .filter(f -> f.dragging.isPresent())
                                    .filter(f -> f.dragging.get() == component)
                                    .findAny().isPresent()) {
                                finger.dragging = Optional.of(component);
                                component.gui()
                                        .sendNewEvent(GuiEvent.PRESS_LEFT,
                                                new GuiComponentEvent(
                                                        finger.cursor.guiX(),
                                                        finger.cursor.guiY()),
                                                component, engine);
                            }
                        });
                fingers.put(tracker, finger);
            } else {
                handleFinger(fetch, ratio);
            }
            newFingers.put(tracker, fetch);
        });
        Streams.of(fingers.keySet())
                .filter(tracker -> !newFingers.containsKey(tracker))
                .forEach(tracker -> {
                    Finger finger = fingers.remove(tracker);
                    if (finger.dragging.isPresent()) {
                        GuiComponent component = finger.dragging.get();
                        if (System.currentTimeMillis() - finger.start < 250) {
                            component.gui().sendNewEvent(GuiEvent.CLICK_LEFT,
                                    new GuiComponentEvent(finger.cursor.guiX(),
                                            finger.cursor.guiY()), component,
                                    engine);
                        }
                        component.gui().sendNewEvent(GuiEvent.DROP_LEFT,
                                new GuiComponentEvent(finger.cursor.guiX(),
                                        finger.cursor.guiY()), component,
                                engine);
                    }
                });
        clicks = newClicks;
    }

    @Override
    public void focusTextField(TextFieldData data, boolean multiline) {
        engine.container().dialog("Input", data, multiline);
    }

    @Override
    public boolean processTextField(TextFieldData data, boolean multiline) {
        return true;
    }

    @Override
    public Stream<GuiCursor> cursors() {
        return Streams.of(fingers.values()).map(Finger::cursor);
    }

    @Override
    public Stream<Pair<GuiCursor, ControllerBasic.PressEvent>> clicks() {
        return Streams.of(clicks);
    }

    @Override
    public boolean captureCursor() {
        return false;
    }

    private void handleFinger(Finger finger, double ratio) {
        finger.cursor.set(finger.tracker.now(),
                finger.tracker.now().multiply(ratio));
        if (finger.dragging.isPresent()) {
            GuiComponent component = finger.dragging.get();
            double relativeX = finger.cursor.guiX() - finger.dragX;
            double relativeY = finger.cursor.guiY() - finger.dragY;
            finger.dragX = finger.cursor.guiX();
            finger.dragY = finger.cursor.guiY();
            component.gui().sendNewEvent(GuiEvent.DRAG_LEFT,
                    new GuiComponentEvent(finger.cursor.guiX(),
                            finger.cursor.guiY(), relativeX, relativeY),
                    component, engine);
            Vector2 source = finger.source.multiply(ratio);
            engine.guiStack().fireRecursiveEvent(GuiEvent.SCROLL,
                    new GuiComponentEvent(source.doubleX(), source.doubleY(),
                            relativeX, relativeY), engine);
        }
    }

    private static final class Finger {
        private final MutableVector2 tracker;
        private final Vector2 source;
        private final long start;
        private final GuiCursor cursor = new GuiCursor();
        private Optional<GuiComponent> dragging = Optional.empty();
        private double dragX, dragY;

        private Finger(MutableVector2 tracker) {
            this.tracker = tracker;
            source = tracker.now();
            start = System.currentTimeMillis();
        }

        private GuiCursor cursor() {
            return cursor;
        }
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/client/input/touch/GuiControllerTouch.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.client.input.touch

import org.tobi29.scapes.engine.ScapesEngine
import org.tobi29.scapes.engine.gui.*
import org.tobi29.scapes.engine.input.ControllerBasic
import org.tobi29.scapes.engine.input.ControllerTouch
import org.tobi29.scapes.engine.utils.ConcurrentHashMap
import org.tobi29.scapes.engine.utils.math.vector.MutableVector2d

class GuiControllerTouch(engine: ScapesEngine,
                         private val controller: ControllerTouch) : GuiController(
        engine) {
    private val fingers = ConcurrentHashMap<ControllerTouch.Tracker, Finger>()
    private var clicks: List<Pair<GuiCursor, ControllerBasic.PressEvent>> = emptyList()

    override fun update(delta: Double) {
        val newClicks = ArrayList<Pair<GuiCursor, ControllerBasic.PressEvent>>()
        val newFingers = ConcurrentHashMap<ControllerTouch.Tracker, Finger>()
        controller.fingers().forEach { tracker ->
            var fetch: Finger? = fingers[tracker]
            if (fetch == null) {
                fetch = Finger(tracker.pos)
                val finger = fetch
                handleFinger(finger)
                val guiPos = finger.cursor.currentPos()
                finger.dragX = guiPos.x
                finger.dragY = guiPos.y
                engine.guiStack.fireEvent(GuiComponentEvent(guiPos.x, guiPos.y),
                        { _, _ -> true })?.let { component ->
                    if (!fingers.values.any { it.dragging == component }) {
                        finger.dragging = component
                        component.gui.sendNewEvent(GuiEvent.PRESS_LEFT,
                                GuiComponentEvent(guiPos.x, guiPos.y),
                                component)
                    }
                }
                fingers.put(tracker, finger)
            } else {
                handleFinger(fetch)
            }
            newFingers.put(tracker, fetch)
        }
        fingers.keys.asSequence().filter {
            !newFingers.containsKey(it)
        }.forEach { tracker ->
            fingers.remove(tracker)?.let { finger ->
                finger.dragging?.let { component ->
                    val guiPos = finger.cursor.currentPos()
                    if (!finger.clicked) {
                        finger.clicked = true
                        component.gui.sendNewEvent(GuiEvent.CLICK_LEFT,
                                GuiComponentEvent(guiPos.x, guiPos.y),
                                component)
                    }
                    component.gui.sendNewEvent(GuiEvent.DROP_LEFT,
                            GuiComponentEvent(guiPos.x, guiPos.y), component)
                }
            }
        }
        clicks = newClicks
    }

    override fun focusTextField(data: GuiController.TextFieldData,
                                multiline: Boolean) {
        engine.container.dialog("Input", data, multiline)
    }

    override fun processTextField(data: GuiController.TextFieldData,
                                  multiline: Boolean): Boolean {
        return true
    }

    override fun cursors(): Sequence<GuiCursor> {
        return fingers.values.asSequence().map { it.cursor }
    }

    override fun clicks(): Sequence<Pair<GuiCursor, ControllerBasic.PressEvent>> {
        return clicks.asSequence()
    }

    override fun captureCursor(): Boolean {
        return false
    }

    private fun handleFinger(finger: Finger) {
        finger.cursor.set(finger.tracker.now())
        finger.dragging?.let { component ->
            val guiPos = finger.cursor.currentPos()
            val relativeX = guiPos.x - finger.dragX
            val relativeY = guiPos.y - finger.dragY
            finger.dragX = guiPos.x
            finger.dragY = guiPos.y
            component.gui.sendNewEvent(GuiEvent.DRAG_LEFT,
                    GuiComponentEvent(guiPos.x, guiPos.y, relativeX, relativeY),
                    component)
            val source = finger.source
            engine.guiStack.fireRecursiveEvent(GuiEvent.SCROLL,
                    GuiComponentEvent(source.x, source.y, relativeX, relativeY))
            if (System.nanoTime() - finger.start >= 250000000L && !finger.clicked) {
                finger.clicked = true
                finger.dragging?.let { component ->
                    component.gui.sendNewEvent(GuiEvent.CLICK_RIGHT,
                            GuiComponentEvent(guiPos.x, guiPos.y), component)
                }
            }
        }
    }

    private class Finger(val tracker: MutableVector2d) {
        var source = tracker.now()
        var start = System.nanoTime()
        var cursor = GuiCursor()
        var dragging: GuiComponent? = null
        var dragX: Double = 0.0
        var dragY: Double = 0.0
        var clicked = false
    }
}
